File: #1_Two_Sum/Solution.java
/* A hash table based O(n) solution */

import java.util.*;

public class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<Integer, Integer>();
		int[] result = new int[2];
        for (int i = 0; i < nums.length; i++) {
        	if (map.containsKey(target - nums[i])) {
        		result[0] = map.get(target - nums[i]) + 1;
        		result[1] = i + 1;
        		break;
        	}
        	map.put(nums[i], i);
        }
        return result;
    }

    public static void main(String[] args) {
    	int[] t = {2, 7, 11, 15};
    	int target = 9;
    	Solution sol = new Solution();
    	int[] res = sol.twoSum(t, target);
    	System.out.println(res[0] + ", " + res[1]);
    }
}

File: #1_Two_Sum/Solution.py
'''
Two pointer Solution O(nlog(n))

Since the question requires to return index, we need to use additional space to save position information
Not best fit for this problem
'''

class Num:
	def __init__(self, index, val):
		self.index = index
		self.val = val

class Solution:
    # @param {integer[]} nums
    # @param {integer} target
    # @return {integer[]}
    def twoSum(self, nums, target):
    	# save index info
    	NumArray = [Num(x, nums[x]) for x in range(len(nums))]
    	# sort the array
    	NumArray.sort(key = lambda Num: Num.val)
    	# use two pointers
    	i, j = 0, len(NumArray) - 1
    	# see if they add up to target val
    	while NumArray[i].val + NumArray[j].val != target:
    		if NumArray[i].val + NumArray[j].val > target:
    			j -= 1
    		if NumArray[i].val + NumArray[j].val < target:
    			i += 1
    	re = [NumArray[i].index + 1, NumArray[j].index + 1]
    	return sorted(re)

nums = [3, 2, 4]
sol = Solution()
print sol.twoSum(nums, 6)